python/taichi/ui/camera.py

```diff
@@ -217,50 +221,49 @@ def track_user_inputs(self,
         position_change = Vector([0.0, 0.0, 0.0])
         left = self.curr_up.cross(front)
         up = self.curr_up
-        if window.is_pressed('w'):
-            position_change = front * movement_speed
-        if window.is_pressed('s'):
-            position_change = -front * movement_speed
-        if window.is_pressed('a'):
-            position_change = left * movement_speed
-        if window.is_pressed('d'):
-            position_change = -left * movement_speed
-        if window.is_pressed('e'):
-            position_change = up * movement_speed
-        if window.is_pressed('q'):
-            position_change = -up * movement_speed
+
+        if self.last_time is None:
+            self.last_time = time.perf_counter_ns()
+        time_elapsed = (time.perf_counter_ns() - self.last_time) * 1e-9
+        self.last_time = time.perf_counter_ns()
+
+        movement_speed *= time_elapsed
+        if window.is_pressed("w"):
+            position_change += front * movement_speed
+        if window.is_pressed("s"):
+            position_change -= front * movement_speed
+        if window.is_pressed("a"):
+            position_change += left * movement_speed
+        if window.is_pressed("d"):
+            position_change -= left * movement_speed
+        if window.is_pressed("e"):
+            position_change += up * movement_speed
+        if window.is_pressed("q"):
+            position_change -= up * movement_speed
         self.position(*(self.curr_position + position_change))
         self.lookat(*(self.curr_lookat + position_change))
 
-        if hold_key is not None:
-            if not window.is_pressed(hold_key):
-                self.last_mouse_x = None
-                self.last_mouse_y = None
-                return
-
         curr_mouse_x, curr_mouse_y = window.get_cursor_pos()
-        if self.last_mouse_x is None or self.last_mouse_y is None:
-            self.last_mouse_x, self.last_mouse_y = curr_mouse_x, curr_mouse_y
-            return
-
-        dx = curr_mouse_x - self.last_mouse_x
-        dy = curr_mouse_y - self.last_mouse_y
-
-        yaw, pitch = vec_to_euler(front)
-        yaw_speed = 2
-        pitch_speed = 2
-
-        yaw -= dx * yaw_speed
-        pitch += dy * pitch_speed
-
-        pitch_limit = pi / 2 * 0.99
-        if pitch > pitch_limit:
-            pitch = pitch_limit
-        elif pitch < -pitch_limit:
-            pitch = -pitch_limit
-
-        front = euler_to_vec(yaw, pitch)
-        self.lookat(*(self.curr_position + front))
-        self.up(0, 1, 0)
+        if (hold_key is None) or window.is_pressed(hold_key):
+            if (self.last_mouse_x is None) or (self.last_mouse_y is None):
+                self.last_mouse_x, self.last_mouse_y = curr_mouse_x, curr_mouse_y
+            dx = curr_mouse_x - self.last_mouse_x
+            dy = curr_mouse_y - self.last_mouse_y
+
+            yaw, pitch = vec_to_euler(front)
+
+            yaw_speed *= time_elapsed
```

Comment:
```suggestion
            yaw_speed *= time_elapsed * 60.0f
```

So that we match the speed assuming most people are using 60hz vsync

Reply:
How about adding an optional parameter `fps: float = 60.0` for more flexibility?

Reply:
I have tested & committed the change. Please let me know if I made anything wrong.

Reply:
Let's just hard code the value, since the time independent update is actually the better way, this is just to ensure it's mostly similar the experience of before

Reply:
👌🏻 I have modified the last commit.